src/Http/Routing/src/RouteEndpointBuilder.cs

```diff
@@ -79,9 +80,50 @@ public override Endpoint Build()
             requestDelegate,
             RoutePattern,
             Order,
-            new EndpointMetadataCollection(Metadata),
+            CreateMetadataCollection(Metadata),
             DisplayName);
 
         return routeEndpoint;
     }
+
+    private static EndpointMetadataCollection CreateMetadataCollection(IList<object> metadata)
+    {
+        if (metadata is { Count: > 0 })
+        {
+            var hasCorsMetadata = false;
+            IHttpMethodMetadata? httpMethodMetadata = null;
+
+            // Before create the final collection we
+            // need to update the IHttpMethodMetadata if
+            // a CORS metadata is present
+            for (var i = 0; i < metadata.Count; i++)
+            {
+                // Not using else if since a metadata could have both
+                // interfaces.
+
+                if (metadata[i] is IHttpMethodMetadata methodMetadata)
+                {
+                    // Storing only the last entry
+                    // since the last metadata is the most significant.
+                    httpMethodMetadata = methodMetadata;
+                }
+
+                if (!hasCorsMetadata && metadata[i] is ICorsMetadata)
+                {
+                    // IEnableCorsAttribute, IDisableCorsAttribute and ICorsPolicyMetadata
+                    // are ICorsMetadata
+                    hasCorsMetadata = true;
+                }
+            }
+
+            if (hasCorsMetadata && httpMethodMetadata is not null)
```

Comment:
Should we also try to avoid potential issues if `AcceptCorsPreflight` is already true? If we do this, we should also add a regression test for the behavior.

```suggestion
            if (hasCorsMetadata && httpMethodMetadata is not null && !httpMethodMetadata.AcceptCorsPreflight)
```